pkg/pushsync/pushsync.go

```diff
@@ -3,3 +3,215 @@
+	go ps.chunksWorker(ctx)
+
+	return ps
+}
+
+func (s *PushSync) Protocol() p2p.ProtocolSpec {
+	return p2p.ProtocolSpec{
+		Name:    protocolName,
+		Version: protocolVersion,
+		StreamSpecs: []p2p.StreamSpec{
+			{
+				Name:    streamName,
+				Handler: s.handler,
+			},
+		},
+	}
+}
+
+func (ps *PushSync) Close() error {
+	close(ps.quit)
+	return nil
+}
+
+// handler handles chunk delivery from other node and inserts it to localstore.
+// it also sends this chunk to the closest peer if one exists.
+func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
+	_, r := protobuf.NewWriterAndReader(stream)
+	defer stream.Close()
+
+	var ch pb.Delivery
+
+	if err := r.ReadMsg(&ch); err != nil {
+		if err == io.EOF {
+			return nil
+		}
+		return err
+	}
+
+	// create chunk and store it in the local store
+	addr := swarm.NewAddress(ch.Address)
+	chunk := swarm.NewChunk(addr, ch.Data)
+	_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+	if err != nil {
+		return err
+	}
+
+	// push this to the closest node too
+	peer, err := ps.peerSuggester.ClosestPeer(addr)
+	if err != nil {
+		if errors.Is(err, topology.ErrWantSelf) {
+			// i'm the closest - nothing to do
+			return nil
+		}
+		return err
+	}
+	if err := ps.sendChunkMsg(ctx, peer, chunk); err != nil {
+		ps.metrics.SendChunkErrorCounter.Inc()
+		return err
+	}
+
+	return ps.storer.Set(ctx, storage.ModeSetSyncPush, chunk.Address())
+}
+
+// chunksWorker polls localstore sends chunks to peers.
+func (ps *PushSync) chunksWorker(ctx context.Context) {
+	var chunks <-chan swarm.Chunk
+	var unsubscribe func()
+
+	// timer, initially set to 0 to fall through select case on timer.C for initialisation
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	chunksInBatch := -1
+
+	for {
+		select {
+		// handle incoming chunks
+		case ch, more := <-chunks:
+			// if no more, set to nil, reset timer to 0 to finalise batch immediately
+			if !more {
+				chunks = nil
+				var dur time.Duration
+				if chunksInBatch == 0 {
+					dur = 500 * time.Millisecond
+				}
+				timer.Reset(dur)
+				break
+			}
+
+			chunksInBatch++
+			ps.metrics.SendChunkCounter.Inc()
+			peer, err := ps.peerSuggester.ClosestPeer(ch.Address())
+			if err != nil {
+				if errors.Is(err, topology.ErrWantSelf) {
+					if err := ps.storer.Set(ctx, storage.ModeSetSyncPush, ch.Address()); err != nil {
+						ps.logger.Error("pushsync: error setting chunks to synced", "err", err)
+					}
+					continue
+				}
+			}
+
+			if err := ps.sendChunkMsg(ctx, peer, ch); err != nil {
+				ps.metrics.SendChunkErrorCounter.Inc()
+				ps.logger.Errorf("error sending chunk", "addr", ch.Address().String(), "err", err)
+			}
+
+			// set chunk status to synced, insert to db GC index
+			if err := ps.storer.Set(ctx, storage.ModeSetSyncPush, ch.Address()); err != nil {
+				ps.logger.Error("pushsync: error setting chunks to synced", "err", err)
+			}
+
+			// retry interval timer triggers starting from new
+		case <-timer.C:
```

Comment:
also, another thought is that this ugliness could be removed by making `SubscribePush` to just iterate until its empty. I think that this timer does not necessarily improve anything, but just slows things down. We have no proof that this actually makes things better. I think that infinitely iterating on push index will be nicer, rate limiting can be done implicitly by the length of the time it takes for the handler here to handle each chunk that is sent (i.e. if it is a blocking or an asynchronous operation)
cc @janos

Reply:
I agree that this is a bit strange. Even if there are chunks in from `chunks` channel it is possible to get to this case when the timer fires and resets the subscription. The reason to retry would be if there are chunks put in the push index at earlier store time then the chunks that we receive from the subscription. I do not think that this could happen. I think that this is a good suggestion. 👍

Reply:
> The reason to retry would be if there are chunks put in the push index at earlier store time then the chunks that we receive from the subscription

Then why shouldn't the subscription just iterate over all items in the index? this has to be done anyway to flush it empty no?

Reply:
It does the iterate over all items, but iteration is only in forward.